### PR TITLE
add LogMelFeatureExtraction

### PR DESCRIPTION
**Pull request by JackTemaki**
Adds PyTorch-based log-mel feature extraction that is compatible to the librosa-based feature extraction in RETURNN.

The log-scaling and normalization are missing on purpose, as this should be an independent user decision.

Is a draft for now as test cases are still missing, but feel free to comment!

### CODE REVIEW

File: i6_models/primitives/feature_extraction.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+@dataclass
+class LogMelFeatureExtractionV1Config(ModelConfiguration):
+    """
+    Attributes:
+        sample_rate: audio sample rate in Hz
+        win_size: window size in seconds
+        hop_size: window shift in seconds
+        f_min: minimum mel filter frequency in Hz
+        f_max: maximum mel fitler frequency in Hz
```
Comment (vieting) [inline i6_models/primitives/feature_extraction.py:21]:
```suggestion
        f_max: maximum mel filter frequency in Hz
```

Comment (vieting) [reply i6_models/primitives/feature_extraction.py:21]: Also, I think I'd remove the "mel" from this line and the one above.

File: i6_models/primitives/feature_extraction.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+        :param raw_audio: [B, T]
+        :param length in samples: [B]
+        :return features as [B,T,F] and length in frames [B]
+        """
+
+        S = (
+            torch.abs(
+                torch.stft(
+                    raw_audio,
+                    n_fft=self.n_fft,
```
Comment (vieting) [inline i6_models/primitives/feature_extraction.py:83]: This does not allow a window size different from the FFT size. I think we do use that in other setups.

File: i6_models/primitives/feature_extraction.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        self.register_buffer("window", torch.hann_window(int(cfg.win_size * cfg.sample_rate)))
+
+    def forward(self, raw_audio, length) -> Tuple[torch.Tensor, torch.Tensor]:
+        """
+        :param raw_audio: [B, T]
+        :param length in samples: [B]
+        :return features as [B,T,F] and length in frames [B]
+        """
+
+        S = (
```
Comment (vieting) [inline i6_models/primitives/feature_extraction.py:79]: Is there a reason for choosing this variable name?

**Review by vieting — COMMENTED** [review]
Looks mostly good to me, just some small comments.

File: i6_models/primitives/feature_extraction.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+@dataclass
+class LogMelFeatureExtractionV1Config(ModelConfiguration):
+    """
+    Attributes:
+        sample_rate: audio sample rate in Hz
+        win_size: window size in seconds
+        hop_size: window shift in seconds
+        f_min: minimum mel filter frequency in Hz
+        f_max: maximum mel filter frequency in Hz
```
Comment (JackTemaki) [inline i6_models/primitives/feature_extraction.py:21]:
```suggestion
        f_min: minimum filter frequency in Hz
        f_max: maximum filter frequency in Hz
```

File: i6_models/primitives/feature_extraction.py (hunk 2 above)
Comment (JackTemaki) [reply i6_models/primitives/feature_extraction.py:83]: For RETURNN feature extraction, this was never possible! But sure, I can add that.

File: i6_models/primitives/feature_extraction.py (hunk 3 above)
Comment (JackTemaki) [reply i6_models/primitives/feature_extraction.py:79]: will change

File: i6_models/primitives/feature_extraction.py (hunk 2 above)
Comment (vieting) [reply i6_models/primitives/feature_extraction.py:83]: For log Mel features in the network, this was possible because you can specify it directly for the STFT.

Comment (JackTemaki) [reply i6_models/primitives/feature_extraction.py:83]: Yes sure, I was talking about the "old" librosa feature extraction that I am copying here.

File: tests/test_feature_extraction.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        min_amp=1e-10,
+        num_filters=80,
+        center=True,
+    )
+    fe_center = LogMelFeatureExtractionV1(cfg=fe_center_cfg)
+    fe_no_center_cfg = copy.deepcopy(fe_center_cfg)
+    fe_no_center_cfg.center = False
+    fe_no_center = LogMelFeatureExtractionV1(cfg=fe_no_center_cfg)
+    for i in range(10):
+        audio_length = int(numpy.random.randint(10000, 50000))
```
Comment (vieting) [inline tests/test_feature_extraction.py:59]: Should the random seed be fixed in order to make the test deterministic?

**Review by vieting — CHANGES_REQUESTED** [review]
Some more small comments.

File: tests/test_feature_extraction.py (hunk 5 above)
Comment (JackTemaki) [reply tests/test_feature_extraction.py:59]: There is no problem in having it non-deterministic, it could even catch an edge case this way (which should not exist though)

File: i6_models/primitives/feature_extraction.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+
+    def __post_init__(self) -> None:
+        super().__post_init__()
+        assert self.f_max <= self.sample_rate // 2, "f_max can not be larger than half of the sample rate"
+        assert self.f_min > 0 and self.f_max > 0 and self.sample_rate > 0, "frequencies need to be positive"
+        assert self.win_size > 0 and self.hop_size > 0, "window settings need to be positive"
+        assert self.num_filters > 0, "number of filters needs to be positive"
+        assert self.hop_size <= self.win_size, "using a larger hop size than window size does not make sense"
+        assert (
+            self.n_fft is None or self.n_fft >= self.win_size * self.sample_rate
```
[the remark this answers is not in the document]
Comment (JackTemaki) [reply i6_models/primitives/feature_extraction.py]: The int conversion is not needed here

**Review by vieting — APPROVED** [review]

**Comment by JackTemaki** [comment]
@mmz33 as this is relevant for you and Atanas, can you review?
